### ISSUE TITLE
App shell HTML content is not created

### ISSUE THREAD

**Issue by ghost** [DockYard/ember-app-shell#13]
Using the latest (`0.4.1`) version on a new ember (`2.16`) app with `ember-export-application-global` installed.

It looks like:
```
.then(() => Runtime.evaluate({ awaitPromise: true, expression: `
  ${this._appGlobal()}.visit('${visitPath}')
    .then((application) => {
      return document.body.querySelector('.ember-view').outerHTML;
    });
`}))
```
is actually "erroring silently" with the following output:
```
{ 
Error: Result of the evaluation is not a promise
    at /Users/ar/Projects/temp/ember-app-shell/node_modules/chrome-remote-interface/lib/chrome.js:91:30
    at Chrome.handleMessage (/Users/ar/Projects/temp/ember-app-shell/node_modules/chrome-remote-interface/lib/chrome.js:292:13) 
... 
}
```

When wrapping `${this._appGlobal()}.visit('${visitPath}')` inside a native promise it looks like that evaluation is indeed not a promise.

```
return Runtime.evaluate({ awaitPromise: true, expression: `
  new Promise((resolve, reject) => {
    window['${this._appGlobal()}'].visit('/app-shell')
      .then((successResult) => resolve('succes'), (errorResult) => reject(errorResult.toString()))
      .catch((e) => reject('catch', e));
  }).then((suc) => 'suc', (err) => err);
`})
```

Returns:
```
{ result:
   { type: 'string',
     value: 'Error: Assertion Failed: You cannot use the same root element (BODY) multiple times in an Ember.Application' } }
```
which is an EmberError instead. Which makes sense since the app is already booted up:
`window['${this._appGlobal()}']['autoboot'];` -> `{ result: { type: 'boolean', value: true } }`
`window['${this._appGlobal()}']['_booted'];` -> `{ result: { type: 'boolean', value: true } }`

Removing the `${this._appGlobal()}.visit('${visitPath}')` entirely and using only `document.body.querySelector('.ember-view').outerHTML` instead solves the issue for me.

I understand that it was added due to app not being ready on load:
https://github.com/DockYard/ember-app-shell/issues/6

So I'm a little confused why it's not working as expected for me.
![mnki](https://user-images.githubusercontent.com/1056569/32691330-a1ef380e-c70e-11e7-9962-41f5411685af.gif)

**Comment by bcardarella** [comment]
Are you in prod and not exporting the global for that env?

**Comment by ghost** [comment]
I'm in prod and the application global *is exported* as per README instructions.

Created a minimum viable repo for the issue I'm experiencing: 
https://github.com/erkarl/ember-app-shell-test/commit/773fca4365ab48b3ab223d1545fa37b62744e8ff

Running `ember build --prod` and logging out the value of `window['${this._appGlobal()}']['_booted']` 
 within `Runtime.evaluate` sometimes returns me either:
`{ result: { type: 'boolean', value: true } }`
OR
`... description: 'TypeError: Cannot read property \'_booted\' of undefined\n    at <anonymous>:2:36', ...`

Which makes be believe there's a race condition between the app booting up (since EmberApp's `autoboot` is `true`) and hitting the `AppGlobal.visit` API.

As a result when we try to hit the application global's visit the following uncaught error occurs:
`{ Error: Result of the evaluation is not a promise }`

Since the application is already booted up the `AppGlobal.visit` will return an EmberError instead of a Promise: `'Error: Assertion Failed: You cannot use the same root element (BODY) multiple times in an Ember.Application'`.

One possible solution would be to wrap the visit call within a Promise to catch the error and default to still getting the outerHTML:
https://github.com/erkarl/ember-app-shell/commit/db3a61519bea0979e914f0f5c0a0b0958c55f53b

**Comment by bcardarella** [comment]
Reading back, I believe the issue is that the app is booting faster than the app shell is able to redirect via the visit API. I suspect we'll need to set the `autoboot` to false to prevent the app from booting too soon. This should prevent multiple app renders.

I won't be able to get to this until later this week but if you wanted to give this hypothesis a try you can use the Page.addScriptToEvaluateOnLoad function: https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-addScriptToEvaluateOnLoad which will allow you to inject a script after the page load has completed, but before any other JS has executed. For the `scriptSource` you should use:

```js
scriptSource: `${this._appGlobal()}.autoboot = false`
```

That should force the application to not boot.

**Comment by ghost** [comment]
I took some time to play around with `addScriptToEvaluateOnLoad` & `addScriptToEvaluateOnNewDocument` and unfortunately `${this._appGlobal()}.autoboot` gets overwritten to true when the application JS executes.

Additionally I'm also getting mixed results when evaluating `appGlobal.autoboot`:
![screen shot 2017-11-12 at 22 05 10](https://user-images.githubusercontent.com/1056569/32702884-14622eae-c7f6-11e7-9a1a-f6890ec8c9e9.png)

**Comment by bcardarella** [comment]
@erkarl how about using the `application` event?

```js
window.addEventListener('application', ({ detail: Application }) => {
  Application.autoboot = false;
});
```

**Comment by ghost** [comment]
Thanks, using the `application` event works well :).

This, however, didn't solve my 2nd issue which was `appGlobal` being defined consistently within `Runtime.evaluate`.

My suggested solution for this is the following which produces me constant results:
https://github.com/DockYard/ember-app-shell/pull/14

**Comment by ghost** [comment]
Fixed in `0.5.0`

**Comment by ghost** [comment]
Thanks. :)
